tests/python/test_python_frontend.py

```diff
@@ -4344,5 +4344,38 @@ def fusion_func(fd: FusionDefinition):
         nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
 
 
+    def test_reshape_dynamic(self):
+        inputs = [
+            32,
+            torch.randn((192,), dtype=torch.float32, device='cuda:0').as_strided((4, 8, 6), (48, 6, 1)),
+        ]
+
+        def fusion_func(fd : FusionDefinition) -> None :
+            S0 = fd.define_scalar(None, dtype=DataType.Int)
+            T1 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.Float, is_cpu=False, stride_order=[2, 1, 0])
+            S2 = fd.define_scalar(1, dtype=DataType.Int)
+            S3 = fd.ops.mul(S2, S0)
+            S4 = fd.ops.signbit(S3)
+            S5 = fd.define_scalar(False, dtype=DataType.Bool)
+            S6 = fd.ops.ne(S4, S5)
+            S7 = fd.define_scalar(192, dtype=DataType.Int)
+            S8 = fd.ops.fmod(S7, S3)
+            S9 = fd.ops.cast(S8, dtype=DataType.Int)
+            S10 = fd.define_scalar(0, dtype=DataType.Int)
+            S11 = fd.ops.ne(S9, S10)
+            S12 = fd.ops.bitwise_and(S6, S11)
+            S13 = fd.define_scalar(192, dtype=DataType.Int)
+            S14 = fd.ops.reciprocal(S3)
+            S15 = fd.ops.mul(S13, S14)
+            S16 = fd.ops.cast(S12, dtype=DataType.Int)
+            S17 = fd.ops.sub(S15, S16)
+            V18 = fd.define_vector([S0, S17], dtype=DataType.Int)
+            T19 = fd.ops.reshape(T1, new_shape=V18)
```

Comment:
`reshape` concretization here calls evaluate on the chain of operations that produces V18.

signbit/fmod are the two that is missing from what we support, hence I needed to add those. My naive question is, what's the criteria for adding something to what expr eval?

Reply:
I think we'd like to have evaluate implemented for everything. Note that there is also `NaiveValueMachine::runBinaryOp` and friends in `evaluator_common.cpp`.

Reply:
Looks like I can just add an entry there. But I'm wondering where can I add a test for the entry in `NaiveValueMachine::runBinaryOp`?

Reply:
Good question. If a `PrecomputedValues` is bound to an `ExpressionEvaluator` then you can run `expr_eval.precomputedValues()->evaluate()` as is done here: https://github.com/NVIDIA/Fuser/blob/65799ed56f66b671dfbd51336fa302400e2bdf45/csrc/executor.cpp#L1210. That will run `PrecomputedValues::run()`. It looks like our tests don't actually exercise this very much; the `TernaryOp` test tests that a value can be searched in `PrecomputedValues` but I don't believe that tests the `PrecomputedValues::run()` method: https://github.com/NVIDIA/Fuser/blob/65799ed56f66b671dfbd51336fa302400e2bdf45/tests/cpp/test_evaluator.cpp#L559-L560. That could be added in another PR I think.

Reply:
I'm going to leave `adding tests for PrecomputedValues` in another PR.

But looks like it makes sense to keep the implementation consistent between the two. So I'll at least add r signbit & fmod and cpp tests for that.

Thanks for pointing this one out. :bow:

Reply:
I just realized that `PrecomputedValues` are only used for ID extent() here: https://github.com/NVIDIA/Fuser/blob/6c6f3a40e09e6f8bece80b9b79c543945846c71b/csrc/evaluator_common.cpp#L93-L117

I don't see an easy way to get that with `Signbit` or `fmod`. Would you mind if we leave the testing part out? :stuck_out_tongue_closed_eyes: